test/js-api/wasm-module-builder.js

```diff
@@ -906,24 +902,55 @@ class WasmModuleBuilder {
       });
     }
 
-    return binary;
+    return binary.trunc_buffer();
   }
 
   toBuffer(debug = false) {
-    let bytes = this.toArray(debug);
-    let buffer = new ArrayBuffer(bytes.length);
-    let view = new Uint8Array(buffer);
-    for (let i = 0; i < bytes.length; i++) {
-      let val = bytes[i];
-      if ((typeof val) == "string") val = val.charCodeAt(0);
-      view[i] = val | 0;
-    }
-    return new Uint8Array(buffer);
+    return this.toUint8Array(debug).buffer;
   }
 
-  instantiate(...args) {
-    let module = new WebAssembly.Module(this.toBuffer());
-    let instance = new WebAssembly.Instance(module, ...args);
+  toArray(debug = false) {
+    return Array.from(this.toUint8Array(debug));
+  }
+
+  instantiate(ffi) {
+    let module = this.toModule();
+    let instance = new WebAssembly.Instance(module, ffi);
     return instance;
   }
+
+  asyncInstantiate(ffi) {
+    return WebAssembly.instantiate(this.toBuffer(), ffi)
+        .then(({module, instance}) => instance);
+  }
+
+  toModule(debug = false) {
+    return new WebAssembly.Module(this.toBuffer(debug));
+  }
+}
+
+function wasmI32Const(val) {
+  let bytes = [kExprI32Const];
+  for (let i = 0; i < 4; ++i) {
+    bytes.push(0x80 | ((val >> (7 * i)) & 0x7f));
+  }
+  bytes.push((val >> (7 * 4)) & 0x7f);
+  return bytes;
+}
+
+function wasmF32Const(f) {
+  f32_view[0] = f;
+  return [
+    kExprF32Const, f32_bytes_view[0], f32_bytes_view[1], f32_bytes_view[2],
+    f32_bytes_view[3]
+  ];
+}
+
+function wasmF64Const(f) {
+  f64_view[0] = f;
+  return [
+    kExprF64Const, f64_bytes_view[0], f64_bytes_view[1], f64_bytes_view[2],
+    f64_bytes_view[3], f64_bytes_view[4], f64_bytes_view[5], f64_bytes_view[6],
+    f64_bytes_view[7]
+  return [kExprF64Const, ...f64_bytes_view];
```

Comment:
extra return

Reply:
Argh, too much syncing back and forth here :/
Thanks for catching, fixed.